### PR TITLE
Eng 12307 stopnode bypass partition detection

### PR DESCRIPTION
**Pull request by luyangco**

### CODE REVIEW

File: tests/frontend/org/voltdb/regressionsuites/TestStopNode.java [hunk 1]
```diff
@@ -161,6 +165,99 @@ public void testStopNode() throws Exception {
...
+            client.drain();
+            cdl.await();
+        } catch (Exception pce) {
+            pce.printStackTrace();
+            lostConnect = pce.getMessage().contains("was lost before a response was received");
+        }
+        //We should never lose contact.
+        assertFalse(lostConnect);
+    }
+
```
Comment (xinjiacs) [inline tests/frontend/org/voltdb/regressionsuites/TestStopNode.java]: Do we have any test covering the mix of stop node and node failures? Changing the order of these sequence, what will happen/break?

File: src/frontend/org/voltcore/messaging/ForeignHost.java [hunk 2]
```diff
@@ -433,6 +440,20 @@ public void sendPoisonPill(String err, int cause) {
...
+        if (m_linkCutForTest.get()) {
+            return null;
+        }
+        ByteBuffer message = ByteBuffer.allocate(20);
+        message.putInt(message.capacity() - 4);
+        message.putLong(-1);
+        message.putInt(STOPNODE_NOTICE);
+        message.putInt(targetHostId);
+        message.flip();
+        return m_network.enqueueAndDrain(message);
```
Comment (luyangco) [inline src/frontend/org/voltcore/messaging/ForeignHost.java]: I prefer not to change the name, my reasons are:

First, PicoNetwork already has a function with the same signature `enqueue(ByteBuffer)`, Java compiler would consider it a duplicate method if I create another `enqueue(ByteBuffer)` even with different return type.

Second, it's not just "drain" part but the whole "enqueue and drain" is handled asynchronously, the name reflects this behavior likes other `enqueue()` functions in PicoNetwork.

File: tests/frontend/org/voltdb/regressionsuites/TestStopNode.java (hunk 1 above)
Comment (luyangco) [inline tests/frontend/org/voltdb/regressionsuites/TestStopNode.java]: I'll add a new test case to test mixed behavior.

File: src/frontend/org/voltcore/messaging/HostMessenger.java [hunk 3]
```diff
@@ -1654,6 +1657,35 @@ public void sendPoisonPill(String err, int targetHostId, int cause) {
...
+
+        // Then contact other peers
+        List<FutureTask<Void>> tasks = new ArrayList<FutureTask<Void>>();
+        for (int hostId : m_foreignHosts.keySet()) {
+            if (hostId == m_localHostId) continue; /* skip target host */
+            Iterator<ForeignHost> it = m_foreignHosts.get(hostId).iterator();
+            if (it.hasNext()) {
+                ForeignHost fh = it.next();
+                if (fh.isUp()) {
+                    FutureTask<Void> task = fh.sendStopNodeNotice(targetHostId);
```
Comment (zeemanhe) [inline src/frontend/org/voltcore/messaging/HostMessenger.java]: Will a host be notified multiple times? It shouldn't matter, correct?

Comment (luyangco) [inline src/frontend/org/voltcore/messaging/HostMessenger.java]: Right, this operation is idempotent. It shouldn't matter if any host be notified multiple times with the same targetHostId.

**Comment by xinjiacs** [comment]
Better to add tests for ticket: https://issues.voltdb.com/browse/ENG-11316

Want to confirm that multiple nodes stopped concurrently without any issue triggering the partition detection.

**Comment by luyangco** [comment]
This use case (concurrent stop node) is covered in https://github.com/VoltDB/voltdb/blob/ENG-12307-stopnode-bypas-pd/tests/frontend/org/voltdb/regressionsuites/TestStopNode.java#L168

File: tests/frontend/org/voltdb/regressionsuites/TestStopNode.java [hunk 4]
```diff
@@ -114,70 +114,122 @@ public void waitForHostsToBeGone(Set<Integer> hids, int[] survivorHostIds) {
...
+//            cdl.await();
+//        } catch (Exception pce) {
+//            pce.printStackTrace();
+//            lostConnect = pce.getMessage().contains("was lost before a response was received");
+//        }
+//        //We should never lose contact.
+//        assertFalse(lostConnect);
+//    }
+//
+//    public void testStopThreeNodesSimultaneously() throws Exception {
```
Comment (xinjiacs) [inline tests/frontend/org/voltdb/regressionsuites/TestStopNode.java]: uncomment it?

File: tests/frontend/org/voltdb/regressionsuites/TestStopNode.java [hunk 5]
```diff
@@ -258,51 +266,97 @@ public void testStopNodesMoreThanAllowed() throws Exception {
...
-            cdl.await();
-        } catch (Exception pce) {
-            pce.printStackTrace();
-            lostConnect = pce.getMessage().contains("was lost before a response was received");
-        }
-        //We should never lose contact.
-        assertFalse(lostConnect);
-    }
+//    public void testStopNodesMoreThanAllowed() throws Exception {
+//        if (kfactor < 1) return;
```
Comment (xinjiacs) [inline tests/frontend/org/voltdb/regressionsuites/TestStopNode.java]: accidental changes?

**Comment by luyangco** [comment]
Commented test cases are reverted.

**Pull request merged**